### ISSUE TITLE
Build error on armhfp/neon

### ISSUE THREAD

**Issue by barsnick** [Genivia/ugrep#128]
Hi,
I'm packaging ugrep for Fedora in a so-called [Copr](https://copr.fedorainfracloud.org/), which is a CI service feeding a personal public repository.

Even though I don't use ugrep on the armhfp architecture, I build the packages there anyway. It may be useful for others, and is also a good cross-architecture build check.

Everything is fine, except that, on this particular architecture, I'm getting build errors from what are probably intrinsics, e.g.
```
g++ -std=gnu++11 -DHAVE_CONFIG_H -I. -I..  -I../include -march=native -mfpu=neon -DHAVE_NEON -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include  -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -march=armv7-a -mfpu=vfpv3-d16 -mtune=generic-armv7-a -mabi=aapcs-linux -mfloat-abi=hard -c -o libreflex_a-matcher.o `test -f 'matcher.cpp' || echo './'`matcher.cpp
In file included from ../include/reflex/input.h:55,
                 from ../include/reflex/absmatcher.h:51,
                 from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
matcher.cpp: In member function 'reflex::Matcher::advance()':
/usr/lib/gcc/armv7hl-redhat-linux-gnueabi/11/include/arm_neon.h:6751:1: error: inlining failed in call to 'always_inline' 'vdupq_n_u8': target specific option mismatch
 6751 | vdupq_n_u8 (uint8_t __a)
      | ^~~~~~~~~~
matcher.cpp:516:35: note: called from here
  516 |       uint8x16_t vlcs = vdupq_n_u8(pre[lcs_]);
      |                         ~~~~~~~~~~^~~~~~~~~~~
```

I have no idea what this is or how to solve it. It may even be a library or compiler issue on this arch, for all I (don't) know. (Other packages build fine on this arch, and ugrep builds fine on other exotic archs such as s390x.)

Compiler versions affected are g++ 10.2.1 (F32), 10.3.1 (F33), 11.0.1 (F34 and Rawhide).

Affected ugrep versions which I have tested are 3.1.9 to 3.1.15.

The configure line is:
```
./configure --build=armv7hl-redhat-linux-gnueabi --host=armv7hl-redhat-linux-gnueabi --program-prefix= --disable-dependency-tracking --prefix=/usr --exec-prefix=/usr --bindir=/usr/bin --sbindir=/usr/sbin --sysconfdir=/etc --datadir=/usr/share --includedir=/usr/include --libdir=/usr/lib --libexecdir=/usr/libexec --localstatedir=/var --sharedstatedir=/var/lib --mandir=/usr/share/man --infodir=/usr/share/info
```
and it shows:
```
checking for --disable-neon... no
checking whether g++ -std=gnu++11 supports ARM NEON/AArch64 intrinsics... yes
```

One complete copr build log [can be seen here](https://download.copr.fedorainfracloud.org/results/barsnick/non-fed/fedora-34-armhfp/02151788-ugrep/builder-live.log.gz), but it may disappear after some time.

<details>
<summary>For the record, here's the complete error output (previous succeeding parts omitted).</summary>
<p>

```
g++ -std=gnu++11 -DHAVE_CONFIG_H -I. -I..  -I../include -march=native -mfpu=neon -DHAVE_NEON -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include  -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -march=armv7-a -mfpu=vfpv3-d16 -mtune=generic-armv7-a -mabi=aapcs-linux -mfloat-abi=hard -c -o libreflex_a-matcher.o `test -f 'matcher.cpp' || echo './'`matcher.cpp
In file included from ../include/reflex/input.h:55,
                 from ../include/reflex/absmatcher.h:51,
                 from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
matcher.cpp: In member function 'reflex::Matcher::advance()':
/usr/lib/gcc/armv7hl-redhat-linux-gnueabi/11/include/arm_neon.h:6751:1: error: inlining failed in call to 'always_inline' 'vdupq_n_u8': target specific option mismatch
 6751 | vdupq_n_u8 (uint8_t __a)
      | ^~~~~~~~~~
matcher.cpp:516:35: note: called from here
  516 |       uint8x16_t vlcs = vdupq_n_u8(pre[lcs_]);
      |                         ~~~~~~~~~~^~~~~~~~~~~
In file included from ../include/reflex/input.h:55,
                 from ../include/reflex/absmatcher.h:51,
                 from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
/usr/lib/gcc/armv7hl-redhat-linux-gnueabi/11/include/arm_neon.h:6751:1: error: inlining failed in call to 'always_inline' 'vdupq_n_u8': target specific option mismatch
 6751 | vdupq_n_u8 (uint8_t __a)
      | ^~~~~~~~~~
matcher.cpp:515:35: note: called from here
  515 |       uint8x16_t vlcp = vdupq_n_u8(pre[lcp_]);
      |                         ~~~~~~~~~~^~~~~~~~~~~
In file included from ../include/reflex/input.h:55,
                 from ../include/reflex/absmatcher.h:51,
                 from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
/usr/lib/gcc/armv7hl-redhat-linux-gnueabi/11/include/arm_neon.h:6328:1: error: inlining failed in call to 'always_inline' 'vgetq_lane_u64': target specific option mismatch
 6328 | vgetq_lane_u64 (uint64x2_t __a, const int __b)
      | ^~~~~~~~~~~~~~
matcher.cpp:525:39: note: called from here
  525 |         uint64_t mask = vgetq_lane_u64(vmask64, 0);
      |                         ~~~~~~~~~~~~~~^~~~~~~~~~~~
In file included from ../include/reflex/input.h:55,
                 from ../include/reflex/absmatcher.h:51,
                 from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
/usr/lib/gcc/armv7hl-redhat-linux-gnueabi/11/include/arm_neon.h:16268:1: error: inlining failed in call to 'always_inline' 'vreinterpretq_u64_u8': target specific option mismatch
16268 | vreinterpretq_u64_u8 (uint8x16_t __a)
      | ^~~~~~~~~~~~~~~~~~~~
matcher.cpp:524:50: note: called from here
  524 |         uint64x2_t vmask64 = vreinterpretq_u64_u8(vmask8);
      |                              ~~~~~~~~~~~~~~~~~~~~^~~~~~~~
In file included from ../include/reflex/input.h:55,
                 from ../include/reflex/absmatcher.h:51,
                 from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
/usr/lib/gcc/armv7hl-redhat-linux-gnueabi/11/include/arm_neon.h:13824:1: error: inlining failed in call to 'always_inline' 'vandq_u8': target specific option mismatch
13824 | vandq_u8 (uint8x16_t __a, uint8x16_t __b)
      | ^~~~~~~~
matcher.cpp:523:37: note: called from here
  523 |         uint8x16_t vmask8 = vandq_u8(vlcpeq, vlcseq);
      |                             ~~~~~~~~^~~~~~~~~~~~~~~~
In file included from ../include/reflex/input.h:55,
                 from ../include/reflex/absmatcher.h:51,
                 from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
/usr/lib/gcc/armv7hl-redhat-linux-gnueabi/11/include/arm_neon.h:2444:1: error: inlining failed in call to 'always_inline' 'vceqq_u8': target specific option mismatch
 2444 | vceqq_u8 (uint8x16_t __a, uint8x16_t __b)
      | ^~~~~~~~
matcher.cpp:522:37: note: called from here
  522 |         uint8x16_t vlcseq = vceqq_u8(vlcs, vlcsm);
      |                             ~~~~~~~~^~~~~~~~~~~~~
In file included from ../include/reflex/input.h:55,
                 from ../include/reflex/absmatcher.h:51,
                 from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
/usr/lib/gcc/armv7hl-redhat-linux-gnueabi/11/include/arm_neon.h:2444:1: error: inlining failed in call to 'always_inline' 'vceqq_u8': target specific option mismatch
 2444 | vceqq_u8 (uint8x16_t __a, uint8x16_t __b)
      | ^~~~~~~~
matcher.cpp:521:37: note: called from here
  521 |         uint8x16_t vlcpeq = vceqq_u8(vlcp, vlcpm);
      |                             ~~~~~~~~^~~~~~~~~~~~~
In file included from ../include/reflex/input.h:55,
                 from ../include/reflex/absmatcher.h:51,
                 from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
/usr/lib/gcc/armv7hl-redhat-linux-gnueabi/11/include/arm_neon.h:10403:1: error: inlining failed in call to 'always_inline' 'vld1q_u8': target specific option mismatch
10403 | vld1q_u8 (const uint8_t * __a)
      | ^~~~~~~~
matcher.cpp:520:36: note: called from here
  520 |         uint8x16_t vlcsm = vld1q_u8(reinterpret_cast<const uint8_t*>(s) + lcs_ - lcp_);
      |                            ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../include/reflex/input.h:55,
                 from ../include/reflex/absmatcher.h:51,
                 from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
/usr/lib/gcc/armv7hl-redhat-linux-gnueabi/11/include/arm_neon.h:10403:1: error: inlining failed in call to 'always_inline' 'vld1q_u8': target specific option mismatch
10403 | vld1q_u8 (const uint8_t * __a)
      | ^~~~~~~~
matcher.cpp:519:36: note: called from here
  519 |         uint8x16_t vlcpm = vld1q_u8(reinterpret_cast<const uint8_t*>(s));
      |                            ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../include/reflex/input.h:55,
                 from ../include/reflex/absmatcher.h:51,
                 from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
/usr/lib/gcc/armv7hl-redhat-linux-gnueabi/11/include/arm_neon.h:6328:1: error: inlining failed in call to 'always_inline' 'vgetq_lane_u64': target specific option mismatch
 6328 | vgetq_lane_u64 (uint64x2_t __a, const int __b)
      | ^~~~~~~~~~~~~~
matcher.cpp:550:30: note: called from here
  550 |         mask = vgetq_lane_u64(vmask64, 1);
      |                ~~~~~~~~~~~~~~^~~~~~~~~~~~
In file included from ../include/reflex/input.h:55,
                 from ../include/reflex/absmatcher.h:51,
                 from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
/usr/lib/gcc/armv7hl-redhat-linux-gnueabi/11/include/arm_neon.h:6751:1: error: inlining failed in call to 'always_inline' 'vdupq_n_u8': target specific option mismatch
 6751 | vdupq_n_u8 (uint8_t __a)
      | ^~~~~~~~~~
matcher.cpp:516:35: note: called from here
  516 |       uint8x16_t vlcs = vdupq_n_u8(pre[lcs_]);
      |                         ~~~~~~~~~~^~~~~~~~~~~
In file included from ../include/reflex/input.h:55,
                 from ../include/reflex/absmatcher.h:51,
                 from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
/usr/lib/gcc/armv7hl-redhat-linux-gnueabi/11/include/arm_neon.h:6751:1: error: inlining failed in call to 'always_inline' 'vdupq_n_u8': target specific option mismatch
 6751 | vdupq_n_u8 (uint8_t __a)
      | ^~~~~~~~~~
matcher.cpp:515:35: note: called from here
  515 |       uint8x16_t vlcp = vdupq_n_u8(pre[lcp_]);
      |                         ~~~~~~~~~~^~~~~~~~~~~
In file included from ../include/reflex/input.h:55,
                 from ../include/reflex/absmatcher.h:51,
                 from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
/usr/lib/gcc/armv7hl-redhat-linux-gnueabi/11/include/arm_neon.h:6328:1: error: inlining failed in call to 'always_inline' 'vgetq_lane_u64': target specific option mismatch
 6328 | vgetq_lane_u64 (uint64x2_t __a, const int __b)
      | ^~~~~~~~~~~~~~
matcher.cpp:525:39: note: called from here
  525 |         uint64_t mask = vgetq_lane_u64(vmask64, 0);
      |                         ~~~~~~~~~~~~~~^~~~~~~~~~~~
In file included from ../include/reflex/input.h:55,
                 from ../include/reflex/absmatcher.h:51,
                 from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
/usr/lib/gcc/armv7hl-redhat-linux-gnueabi/11/include/arm_neon.h:16268:1: error: inlining failed in call to 'make[2]: Leaving directory '/builddir/build/BUILD/ugrep-3.1.15/lib'
always_inline' 'vreinterpretq_u64_u8': target specific option mismatch
16268 | vreinterpretq_u64_u8 (uint8x16_t __a)
      | ^~~~~~~~~~~~~~~~~~~~
matcher.cpp:524:50: note: called from here
  524 |         uint64x2_t vmask64 = vreinterpretq_u64_u8(vmask8);
      |                              ~~~~~~~~~~~~~~~~~~~~^~~~~~~~
In file included from ../include/reflex/input.h:55,
                 from ../include/reflex/absmatcher.h:51,
                 from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
/usr/lib/gcc/armv7hl-redhat-linux-gnueabi/11/include/arm_neon.h:13824:1: error: inlining failed in call to 'always_inline' 'vandq_u8': target specific option mismatch
13824 | vandq_u8 (uint8x16_t __a, uint8x16_t __b)
      | ^~~~~~~~
matcher.cpp:523:37: note: called from here
  523 |         uint8x16_t vmask8 = vandq_u8(vlcpeq, vlcseq);
      |                             ~~~~~~~~^~~~~~~~~~~~~~~~
In file included from ../include/reflex/input.h:55,
                 from ../include/reflex/absmatcher.h:51,
                 from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
/usr/lib/gcc/armv7hl-redhat-linux-gnueabi/11/include/arm_neon.h:2444:1: error: inlining failed in call to 'always_inline' 'vceqq_u8': target specific option mismatch
 2444 | vceqq_u8 (uint8x16_t __a, uint8x16_t __b)
      | ^~~~~~~~
matcher.cpp:522:37: note: called from here
  522 |         uint8x16_t vlcseq = vceqq_u8(vlcs, vlcsm);
      |                             ~~~~~~~~^~~~~~~~~~~~~
In file included from ../include/reflex/input.h:55,
                 from ../include/reflex/absmatcher.h:51,
                 from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
/usr/lib/gcc/armv7hl-redhat-linux-gnueabi/11/include/arm_neon.h:2444:1: error: inlining failed in call to 'always_inline' 'vceqq_u8': target specific option mismatch
 2444 | vceqq_u8 (uint8x16_t __a, uint8x16_t __b)
      | ^~~~~~~~
matcher.cpp:521:37: note: called from here
  521 |         uint8x16_t vlcpeq = vceqq_u8(vlcp, vlcpm);
      |                             ~~~~~~~~^~~~~~~~~~~~~
In file included from ../include/reflex/input.h:55,
                 from ../include/reflex/absmatcher.h:51,
                 from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
/usr/lib/gcc/armv7hl-redhat-linux-gnueabi/11/include/arm_neon.h:10403:1: error: inlining failed in call to 'always_inline' 'vld1q_u8': target specific option mismatch
10403 | vld1q_u8 (const uint8_t * __a)
      | ^~~~~~~~
matcher.cpp:520:36: note: called from here
  520 |         uint8x16_t vlcsm = vld1q_u8(reinterpret_cast<const uint8_t*>(s) + lcs_ - lcp_);
      |                            ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../include/reflex/input.h:55,
                 from ../include/reflex/absmatcher.h:51,
                 from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
/usr/lib/gcc/armv7hl-redhat-linux-gnueabi/11/include/arm_neon.h:10403:1: error: inlining failed in call to 'always_inline' 'vld1q_u8': target specific option mismatch
10403 | vld1q_u8 (const uint8_t * __a)
      | ^~~~~~~~
matcher.cpp:519:36: note: called from here
  519 |         uint8x16_t vlcpm = vld1q_u8(reinterpret_cast<const uint8_t*>(s));
      |                            ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../include/reflex/input.h:55,
                 from ../include/reflex/absmatcher.h:51,
                 from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
/usr/lib/gcc/armv7hl-redhat-linux-gnueabi/11/include/arm_neon.h:6328:1: error: inlining failed in call to 'always_inline' 'vgetq_lane_u64': target specific option mismatch
 6328 | vgetq_lane_u64 (uint64x2_t __a, const int __b)
      | ^~~~~~~~~~~~~~
matcher.cpp:550:30: note: called from here
  550 |         mask = vgetq_lane_u64(vmask64, 1);
      |                ~~~~~~~~~~~~~~^~~~~~~~~~~~
make[2]: *** [Makefile:467: libreflex_a-matcher.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Entering directory '/builddir/build/BUILD/ugrep-3.1.15/lib'
g++ -std=gnu++11 -DHAVE_CONFIG_H -I. -I..  -I../include -march=native -mfpu=neon -DHAVE_NEON -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include  -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -march=armv7-a -mfpu=vfpv3-d16 -mtune=generic-armv7-a -mabi=aapcs-linux -mfloat-abi=hard -c -o libreflex_a-convert.o `test -f 'convert.cpp' || echo './'`convert.cpp
make[2]: Leaving directory '/builddir/build/BUILD/ugrep-3.1.15/lib'
make[1]: *** [Makefile:458: all-recursive] Error 1
make: *** [Makefile:358: all] Error 2
```

</p>
</details>

Cheers!

**Comment by genivia-inc** [comment]
Thank you for the feedback. Will look into this problem soon. I've saved the copr build log to assist with the diagnosis.

**Comment by genivia-inc** [comment]
Looking into the details of the logs, it appears that the Copr scripts include additional compiler flags (not uncommon) with a ARM7a architecture-specific flag `-mfpu=vfpv3-d16` added that overrides `-mfpu=neon`. The `-mfpu=neon` flag was added by the ugrep build scripts, in addition to `-march=native` (which is overridden by Copr by `-march=arm7-a` but that is OK).

There are two possible scenarios: 1) the ugrep `./configure` script wrongly detected the newer NEON architecture of ARM7a after verifying that NEON instructions are supported by compiling some test code, or 2) the Copr scripts always add `-mfpu=vfpv3-d16` for older ARM7 architectures which does not support NEON, even when NEON is actually supported on this machine. If the problem is 1) then we need more testing to see what can be done, but unfortunately this is tricky or may even be impossible to resolve in general, because using `-mfpu=native` in the `./configure` script to check NEON instructions is not an option. If the problem is 2) then I assume the Copr build scripts are not correct, which could be, since your title suggests a **armhfp/neon** machine. Or there may be third scenario, perhaps.

For now, to disable NEON and let Copr build ugrep, run `./build.sh --disable-neon` to work around this.

**Comment by barsnick** [comment]
Good catch on the overwritten "-mfpu=" compile option.

It is definitely scenario 2).

When looking for a way to filter out an automatically set compiler flag (notably in the %configure macro for RPM builds), I noticed that zimg apparently suffered the same issue:
https://src.fedoraproject.org/rpms/zimg/c/4a3cdb03d0d7da25bd4cd92c58040937fc74596b

With that "fix", I managed to build ugrep on these platforms. 💪
https://copr.fedorainfracloud.org/coprs/barsnick/non-fed/build/2160354/

So now I'm building for NEON even though certain ARM7s might not support it? I don't care, and don't know whether those processors are even targetted by this "armhfp" distribution branch:
https://fedoraproject.org/wiki/Architectures/ARM

Thanks for looking at it!
